### DISCUSSION TITLE
Is there any optimisation available?

### DISCUSSION THREAD

**Discussion by ySergi** [lllyasviel/Fooocus#2770] [Q&A] [answered]
In A1111 we can for example use xformers or even TensorRT thanks to an extension, can we do the same here?

**Comment by mashb1t** [comment, marked as answer]
xformers is available, but normally torch >2 is a more performant attention (best variant selected by default).
Please check if setting https://github.com/lllyasviel/Fooocus/blob/e2f9bcb11d06216d6800676c48d8d74d6fd77a4b/launch.py#L26 to true works better for you.
There are also bugs in xformers 0.0.21, 0.0.22 and 0.0.23, see special handling in https://github.com/lllyasviel/Fooocus/blob/e2f9bcb11d06216d6800676c48d8d74d6fd77a4b/ldm_patched/ldm/modules/attention.py#L281C42-L281C110

May I ask which GPU and system specs you have in your system?

**Reply by ySergi** [reply]
I have an 4090 with an i9-13900k and 32GB of RAM.

**Reply by mashb1t** [reply]
@sergiohzph for maximum performance, please check out the options `--always-high-vram --disable-offload-from-vram` or even `--always-gpu`, as xdebug will probably slow you down.

**Reply by ySergi** [reply]
Where do I write those arguments?

**Reply by mashb1t** [reply]
You can put them as argumemts on the entrypoint script. If you're using Windows and have installed Fooocus with the zip file, this would be the the same as editing the .bat scripts used for Fooocus startup.

**Reply by ySergi** [reply]
Well, I've put `--always-high-vram`  in run.bat and it says the following: 
```
Total VRAM 24564 MB, total RAM 32542 MB
Set vram state to: NORMAL_VRAM
```
Shouldn't it be in "HIGH_VRAM" or something like that?

**Reply by mashb1t** [reply]
no, it's normal by default and on startup, then switches when generating afaik
